airflow/models/renderedtifields.py

```diff
@@ -0,0 +1,153 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Save Rendered Template Fields """
+import json
+from typing import Optional
+
+from sqlalchemy import JSON, Column, String, and_, exists, tuple_
+from sqlalchemy.orm import Session
+
+from airflow.configuration import conf
+from airflow.models.base import ID_LEN, Base
+from airflow.models.taskinstance import TaskInstance
+from airflow.serialization.serialization_helpers import serialize_template_field
+from airflow.utils.session import provide_session
+from airflow.utils.sqlalchemy import UtcDateTime
+
+
+class RenderedTaskInstanceFields(Base):
+    """
+    Save Rendered Template Fields
+    """
+
+    __tablename__ = "rendered_task_instance_fields"
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    task_id = Column(String(ID_LEN), primary_key=True)
+    execution_date = Column(UtcDateTime, primary_key=True)
+    rendered_fields = Column(JSON, nullable=False)
+
+    def __init__(self, ti: TaskInstance, render_templates=True):
+        self.dag_id = ti.dag_id
+        self.task_id = ti.task_id
+        self.task = ti.task
+        self.execution_date = ti.execution_date
+        self.ti = ti
+        if render_templates:
+            ti.render_templates()
+        self.rendered_fields = {
+            field: serialize_template_field(
+                getattr(self.task, field)
+            ) for field in self.task.template_fields
+        }
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {self.dag_id}.{self.task_id} {self.execution_date}"
+
+    @classmethod
+    @provide_session
+    def get_templated_fields(cls, ti: TaskInstance, session: Session = None) -> Optional[dict]:
+        """
+        Get templated field for a TaskInstance from the RenderedTaskInstanceFields
+        table.
+
+        :param ti: Task Instance
+        :param session: SqlAlchemy Session
+        :return: Rendered Templated TI field
+        """
+        result = session.query(cls.rendered_fields).filter(
+            cls.dag_id == ti.dag_id,
+            cls.task_id == ti.task_id,
+            cls.execution_date == ti.execution_date
+        ).first()
+
+        if result:
+            rendered_fields = result.rendered_fields
+            if isinstance(rendered_fields, str):
```

Comment:
When would this be a string? Doesn't SQLA automatically parse the JSON for us?

Reply:
It was just a safeguard to DB that doesn't support JSON column type.

I will replace that code with **sqlalchemy_jsonfield** which would handle this for us

Reply:
Updated